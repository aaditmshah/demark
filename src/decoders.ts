import WhiteSpaceCodePoints from "@unicode/unicode-15.0.0/Binary_Property/White_Space/code-points";
import type { Stream, Data, TaggedContent } from "./parser";

const LeftToRightMarkCodePoint = 0x20_0e;
const RightToLeftMarkCodePoint = 0x20_0f;

const WhiteSpaceCharacterCodePoints = [
  ...WhiteSpaceCodePoints,
  LeftToRightMarkCodePoint,
  RightToLeftMarkCodePoint
];

const WhiteSpaceRegExp = new RegExp(
  `^[${WhiteSpaceCharacterCodePoints.map((cp) => String.fromCodePoint(cp)).join(
    ""
  )}]*$`,
  "u"
);

interface State<T> {
  array: T[];
  index: number;
}

interface Result<T, A> {
  state: State<T>;
  value: A;
}

type Decoder<T, A> = (state: State<T>) => Result<T, A>;

type Decoders<T, A> = {
  [K in keyof A]: Decoder<T, A[K]>;
};

const pure =
  <A>(value: A) =>
  <T>(state: State<T>): Result<T, A> => ({ state, value });

const map =
  <T, A, B extends unknown[], C>(
    morphism: (value: A, ...values: B) => C,
    firstDecoder: Decoder<T, A>,
    ...restDecoders: Decoders<T, B>
  ): Decoder<T, C> =>
  (state) => {
    const { state: firstState, value } = firstDecoder(state);
    let currentState = firstState;
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- correct type
    const values = restDecoders.map((decoder) => {
      const result = decoder(currentState);
      currentState = result.state;
      return result.value;
    }) as B;
    return { state: currentState, value: morphism(value, ...values) };
  };

const alt =
  <T, A, B extends Stream<unknown>>(
    firstDecoder: Decoder<T, A>,
    ...restDecoders: Decoders<T, B>
  ): Decoder<T, A | B[number]> =>
  (state) => {
    const errors: unknown[] = [];
    for (const decoder of [firstDecoder, ...restDecoders]) {
      try {
        return decoder(state);
      } catch (error) {
        errors.push(error);
      }
    }
    throw new AggregateError(errors, "all alternatives failed");
  };

const bind =
  <T, A, B>(
    decoder: Decoder<T, A>,
    arrow: (value: A) => Decoder<T, B>
  ): Decoder<T, B> =>
  (state) => {
    const result = decoder(state);
    return arrow(result.value)(result.state);
  };

type List<A> = Cons<A> | null;

interface Cons<A> {
  head: A;
  tail: List<A>;
}

const toArray = <A>(list: List<A>): A[] => {
  const array: A[] = [];
  for (let cursor = list; cursor !== null; cursor = cursor.tail)
    array.push(cursor.head);
  return array;
};

const toStream = <A>(cons: Cons<A>): Stream<A> => {
  const stream: Stream<A> = [cons.head];
  for (let cursor = cons.tail; cursor !== null; cursor = cursor.tail)
    stream.push(cursor.head);
  return stream;
};

interface Kleene<T, A> {
  plus: Decoder<T, Stream<A>>;
  star: Decoder<T, A[]>;
}

const kleene = <T, A>(decoder: Decoder<T, A>): Kleene<T, A> => {
  const some: Decoder<T, Cons<A>> = bind(decoder, (head) =>
    // eslint-disable-next-line @typescript-eslint/no-use-before-define -- mutually recursive
    bind(many, (tail) => pure({ head, tail }))
  );
  // eslint-disable-next-line unicorn/no-null -- empty list
  const many: Decoder<T, List<A>> = alt(some, pure(null));
  return { plus: map(toStream, some), star: map(toArray, many) };
};

const some = <T, A>(decoder: Decoder<T, A>) => kleene(decoder).plus;

const many = <T, A>(decoder: Decoder<T, A>) => kleene(decoder).star;

const at = <T>(array: T[], index: number): T => {
  const element = array.at(index);
  if (typeof element === "undefined")
    throw new Error("unexpected end of input");
  return element;
};

const decode = <T, A>(decoder: Decoder<T, A>, array: T[]) => {
  const { state, value } = decoder({ array, index: 0 });
  if (state.index < array.length) throw new Error("all input not decoded");
  return value;
};

const text: Decoder<Data, string> = ({ array, index }) => {
  const data = at(array, index);
  if (typeof data !== "string") throw new Error("expected text");
  return { state: { array, index: index + 1 }, value: data };
};

const group =
  <A>(decoder: Decoder<TaggedContent, A>): Decoder<Data, A> =>
  ({ array, index }) => {
    const data = at(array, index);
    if (typeof data === "string") {
      if (!WhiteSpaceRegExp.test(data))
        throw new Error("expected tagged content group");
      const nextData = at(array, index + 1);
      // istanbul ignore if -- impossible to test because text never follows text
      if (typeof nextData === "string")
        throw new Error("expected tagged content group");
      return {
        state: { array, index: index + 2 },
        value: decode(decoder, nextData)
      };
    }
    return { state: { array, index: index + 1 }, value: decode(decoder, data) };
  };

const tagged = <A>(
  id: string,
  decoder: Decoder<Data, A>
): Decoder<TaggedContent, A> => {
  const normalizedId = id.normalize("NFC");
  return ({ array, index }) => {
    const { tag, content } = at(array, index);
    if (tag !== normalizedId)
      throw new Error(`expected ${normalizedId} tagged content`);
    const result = decoder({ array: content, index: 0 });
    const remaining = content.length - result.state.index;
    if (remaining > 1) throw new Error("all input not decoded");
    if (remaining === 1) {
      const last = at(content, -1);
      if (typeof last !== "string" || !WhiteSpaceRegExp.test(last))
        throw new Error("all input not decoded");
    }
    return {
      state: { array, index: index + 1 },
      value: result.value
    };
  };
};

export type { State, Result, Decoder, Decoders, Kleene };
export {
  pure,
  map,
  alt,
  bind,
  kleene,
  some,
  many,
  decode,
  text,
  group,
  tagged
};
