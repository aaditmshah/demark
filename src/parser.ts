// eslint-disable-next-line @typescript-eslint/triple-slash-reference -- Ambient Declaration File
/// <reference path="./unicode.d.ts" />
import WhiteSpaceCodePoints from "@unicode/unicode-15.0.0/Binary_Property/White_Space/code-points";
import ControlCodePoints from "@unicode/unicode-15.0.0/General_Category/Control/code-points";
import PrivateUseCodePoints from "@unicode/unicode-15.0.0/General_Category/Private_Use/code-points";
import UnassignedCodePoints from "@unicode/unicode-15.0.0/General_Category/Unassigned/code-points";

interface State {
  bytes: Uint8Array;
  index: number;
}

const NextByte = (state: State) => {
  const nextByte = state.bytes[state.index];
  if (typeof nextByte === "undefined")
    throw new Error("unexpected end of input");
  state.index += 1;
  return nextByte;
};

const ContinuationByte = (state: State) => {
  const continuationByte = NextByte(state);
  if (continuationByte < 0x80) throw new Error("expected continuation byte");
  if (continuationByte < 0xc0) return continuationByte - 0x80;
  if (continuationByte < 0xf5) throw new Error("expected continuation byte");
  throw new Error("invalid byte");
};

const SourceCharacter = (state: State) => {
  const firstByte = NextByte(state);
  if (firstByte < 0x80) return firstByte;
  if (firstByte < 0xc0) throw new Error("unexpected continuation byte");
  if (firstByte < 0xc2) throw new Error("overlong encoding");
  if (firstByte < 0xe0) {
    const bits0 = firstByte - 0xc0;
    const bits1 = ContinuationByte(state);
    return 0x40 * bits0 + bits1;
  }
  if (firstByte < 0xf0) {
    const bits0 = firstByte - 0xe0;
    const bits1 = ContinuationByte(state);
    if (bits0 === 0x0 && bits1 < 0x20) throw new Error("overlong encoding");
    if (bits0 === 0xd && bits1 >= 0x20) throw new Error("surrogate character");
    const bits2 = ContinuationByte(state);
    return 0x10_00 * bits0 + 0x40 * bits1 + bits2;
  }
  if (firstByte < 0xf5) {
    const bits0 = firstByte - 0xf0;
    const bits1 = ContinuationByte(state);
    if (bits0 === 0 && bits1 < 0x10) throw new Error("overlong encoding");
    if (bits0 === 4 && bits1 >= 0x10) throw new Error("invalid code point");
    const bits2 = ContinuationByte(state);
    const bits3 = ContinuationByte(state);
    return 0x4_00_00 * bits0 + 0x10_00 * bits1 + 0x40 * bits2 + bits3;
  }
  throw new Error("invalid byte");
};

const LeftToRightMarkCodePoint = "\u200E".codePointAt(0);
const RightToLeftMarkCodePoint = "\u200F".codePointAt(0);

const WhiteSpaceCharacterCodePoints = new Set([
  ...WhiteSpaceCodePoints,
  LeftToRightMarkCodePoint,
  RightToLeftMarkCodePoint
]);

const NumberSignCodePoint = "#".codePointAt(0);
const ReverseSolidusCodePoint = "\\".codePointAt(0);
const LeftCurlyBracketCodePoint = "{".codePointAt(0);
const RightCurlyBracketCodePoint = "}".codePointAt(0);
const TildeCodePoint = "~".codePointAt(0);

const SyntaxCharacterCodePoints = new Set([
  NumberSignCodePoint,
  ReverseSolidusCodePoint,
  LeftCurlyBracketCodePoint,
  RightCurlyBracketCodePoint,
  TildeCodePoint
]);

const OtherCharacterCodePoints = new Set([
  ...ControlCodePoints,
  ...PrivateUseCodePoints,
  ...UnassignedCodePoints
]);

type Stream<A> = [A, ...A[]];

type Data = string | TaggedContentGroup;

type Content = Data[];

interface TaggedContent {
  tag: string;
  content: Content;
}

type TaggedContentGroup = Stream<TaggedContent>;

const IdentifierCharacter = (codePoint: number) => {
  if (WhiteSpaceCharacterCodePoints.has(codePoint))
    throw new Error("unexpected whitespace character");
  if (SyntaxCharacterCodePoints.has(codePoint))
    throw new Error("unexpected syntax character");
  if (OtherCharacterCodePoints.has(codePoint))
    throw new Error("unexpected other character");
  return String.fromCodePoint(codePoint);
};

const VerbatimContentString = (state: State, level: number) => {
  let string = "";
  while (string.length < level) {
    const codePoint = SourceCharacter(state);
    string += String.fromCodePoint(codePoint);
    if (codePoint !== NumberSignCodePoint) return string;
  }
  const codePoint = SourceCharacter(state);
  string += String.fromCodePoint(codePoint);
  return codePoint === RightCurlyBracketCodePoint ? undefined : string;
};

const VerbatimContent = (state: State): Content => {
  let level = 1;
  let codePoint = SourceCharacter(state);
  while (codePoint !== LeftCurlyBracketCodePoint) {
    if (codePoint !== NumberSignCodePoint)
      throw new Error("expected a left curly bracket");
    level += 1;
    codePoint = SourceCharacter(state);
  }
  let data = "";
  let string = VerbatimContentString(state, level);
  while (typeof string !== "undefined") {
    data += string;
    string = VerbatimContentString(state, level);
  }
  return data === "" ? [] : [data];
};

const NonVerbatimData = (state: State, initialCodePoint: number): Data => {
  let codePoint = initialCodePoint;
  if (codePoint === ReverseSolidusCodePoint) {
    codePoint = SourceCharacter(state);
    return SyntaxCharacterCodePoints.has(codePoint)
      ? String.fromCodePoint(codePoint)
      : [TaggedContentParser(state, codePoint)];
  }
  if (SyntaxCharacterCodePoints.has(codePoint))
    throw new Error("unexpected syntax character");
  return String.fromCodePoint(codePoint);
};

const NonVerbatimContent = (state: State): Content => {
  const content: Content = [];
  let string = "";
  let group: TaggedContentGroup | false = false;
  let codePoint = SourceCharacter(state);
  while (codePoint !== RightCurlyBracketCodePoint) {
    if (group === false) {
      const data = NonVerbatimData(state, codePoint);
      if (typeof data === "string") string += data;
      else {
        if (string !== "") {
          content.push(string);
          string = "";
        }
        content.push(data);
        group = data;
      }
    } else if (WhiteSpaceCharacterCodePoints.has(codePoint)) {
      string += String.fromCodePoint(codePoint);
    } else if (codePoint === TildeCodePoint) {
      group.push(TaggedContentParser(state));
      string = "";
    } else {
      group = false;
      // eslint-disable-next-line no-continue -- intentional
      continue;
    }
    codePoint = SourceCharacter(state);
  }
  return string === "" ? content : [...content, string];
};

function TaggedContentParser(
  state: State,
  initialCodePoint = SourceCharacter(state)
): TaggedContent {
  let codePoint = initialCodePoint;
  let tag = IdentifierCharacter(codePoint);
  codePoint = SourceCharacter(state);
  while (
    !WhiteSpaceCharacterCodePoints.has(codePoint) &&
    codePoint !== LeftCurlyBracketCodePoint &&
    codePoint !== NumberSignCodePoint
  ) {
    if (
      SyntaxCharacterCodePoints.has(codePoint) ||
      OtherCharacterCodePoints.has(codePoint)
    ) {
      throw new Error(
        "expected a whitespace character, a left curly bracket, or a number sign"
      );
    }
    tag += String.fromCodePoint(codePoint);
    codePoint = SourceCharacter(state);
  }
  while (
    codePoint !== LeftCurlyBracketCodePoint &&
    codePoint !== NumberSignCodePoint
  ) {
    if (!WhiteSpaceCharacterCodePoints.has(codePoint))
      throw new Error("expected a left curly bracket or a number sign");
    codePoint = SourceCharacter(state);
  }
  const content =
    codePoint === LeftCurlyBracketCodePoint
      ? NonVerbatimContent(state)
      : VerbatimContent(state);
  return { tag: tag.normalize("NFC"), content };
}

const Document = (state: State) => {
  let codePoint = SourceCharacter(state);
  while (codePoint !== ReverseSolidusCodePoint) {
    if (!WhiteSpaceCharacterCodePoints.has(codePoint))
      throw new Error("expected a reverse solidus");
    codePoint = SourceCharacter(state);
  }
  const taggedContentGroup: TaggedContentGroup = [TaggedContentParser(state)];
  while (state.index < state.bytes.length) {
    codePoint = SourceCharacter(state);
    if (codePoint === TildeCodePoint)
      taggedContentGroup.push(TaggedContentParser(state));
    else if (!WhiteSpaceCharacterCodePoints.has(codePoint))
      throw new Error("expected a tilde");
  }
  return taggedContentGroup;
};

const parse = (bytes: Uint8Array) => Document({ bytes, index: 0 });

export { parse };
