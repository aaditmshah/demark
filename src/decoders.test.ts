import { readFile } from "node:fs/promises";
import { map, alt, some, many, decode, text, group, tagged } from "./decoders";
import type { Stream } from "./parser";
import { parse } from "./parser";

interface Code {
  tag: "code";
  value: string;
}

const createCode = (value: string): Code => ({ tag: "code", value });

const code = group(map(createCode, tagged("code", text)));

interface Link {
  tag: "link";
  value: string | Code;
  href: string;
}

const createLink = (value: string | Code, href: string): Link => ({
  tag: "link",
  value,
  href
});

const link = group(
  map(createLink, tagged("link", alt(text, code)), tagged("href", text))
);

interface Paragraph {
  tag: "paragraph";
  contents: Stream<string | Code | Link>;
}

const createParagraph = (
  contents: Stream<string | Code | Link>
): Paragraph => ({
  tag: "paragraph",
  contents
});

const paragraph = group(
  map(createParagraph, tagged("paragraph", some(alt(text, code, link))))
);

interface Section {
  tag: "section";
  title: string;
  paragraphs: Stream<Paragraph>;
}

const createSection = (
  title: string,
  paragraphs: Stream<Paragraph>
): Section => ({
  tag: "section",
  title,
  paragraphs
});

const section = group(
  map(
    createSection,
    tagged("section", text),
    tagged("content", some(paragraph))
  )
);

interface ArticleContent {
  paragraphs: Stream<Paragraph>;
  sections: Stream<Section>;
}

const createArticleContent = (
  paragraphs: Stream<Paragraph>,
  sections: Stream<Section>
): ArticleContent => ({
  paragraphs,
  sections
});

const articleContent = map(
  createArticleContent,
  some(paragraph),
  some(section)
);

interface Article {
  tag: "article";
  title: string;
  tags: Stream<string>;
  introduction: Stream<Paragraph>;
  body: Stream<Section>;
}

const createArticle = (
  title: string,
  tags: Stream<string>,
  { paragraphs, sections }: ArticleContent
): Article => ({
  tag: "article",
  title,
  tags,
  introduction: paragraphs,
  body: sections
});

const article = map(
  createArticle,
  tagged("article", text),
  some(tagged("tag", text)),
  tagged("content", articleContent)
);

interface CodeLang extends Code {
  language: string;
}

const createCodeLang = (language: string, value: string): CodeLang => ({
  tag: "code",
  language,
  value
});

const codeLang = group(
  map(createCodeLang, tagged("code", text), tagged("content", text))
);

interface Marker {
  tag: "marker";
  value: string;
}

const createMarker = (value: string): Marker => ({ tag: "marker", value });

const marker = group(map(createMarker, tagged("marker", text)));

interface SimpleArticle {
  tag: "article";
  title: string;
  content: (Marker | Code)[];
}

const createSimpleArticle = (
  title: string,
  content: (Marker | Code)[]
): SimpleArticle => ({ tag: "article", title, content });

const simpleArticle = map(
  createSimpleArticle,
  tagged("article", text),
  tagged("content", many(alt(marker, codeLang)))
);

const join2 = (a: string, b: string) => a + b;
const join3 = (a: string, b: string, c: string) => a + b + c;
const join4 = (a: string, b: string, c: string, d: string) => a + b + c + d;

const simpleMarker = group(tagged("marker", text));

const documentError = tagged(
  "foo",
  map(join3, simpleMarker, simpleMarker, text)
);

const documentErrors = alt(
  tagged("foo", map(join4, simpleMarker, simpleMarker, text, simpleMarker)),
  tagged("foo", map(join2, simpleMarker, text)),
  tagged("foo", map(join3, simpleMarker, simpleMarker, simpleMarker)),
  tagged("bar", map(join3, simpleMarker, simpleMarker, text)),
  tagged("foo", simpleMarker),
  tagged("foo", map(join2, simpleMarker, simpleMarker)),
  map(
    join2,
    tagged("foo", map(join3, simpleMarker, simpleMarker, text)),
    tagged("bar", simpleMarker)
  )
);

describe("decode", () => {
  it("should correctly decode parsed Demark documents", async () => {
    expect.assertions(1);
    const bytes = await readFile(
      "./examples/why-create-your-own-eslint-config.dm"
    );
    const result = decode(article, parse(bytes));
    expect(result).toMatchSnapshot();
  });

  it("should correctly decode parsed Demark documents with Unicode", async () => {
    expect.assertions(1);
    const bytes = await readFile("./examples/nested-verbatim-unicode.dm");
    const result = decode(simpleArticle, parse(bytes));
    expect(result).toMatchSnapshot();
  });

  it("should throw an error on decoding invalid Demark documents", async () => {
    expect.assertions(2);
    const bytes = await readFile("./examples/decoder-failure-actual.dm");
    const result = parse(bytes);
    expect(() => decode(documentError, result)).toThrow(
      new Error("all input not decoded")
    );
    expect(() => decode(documentErrors, result)).toThrow(
      new Error("all alternatives failed")
    );
  });
});
