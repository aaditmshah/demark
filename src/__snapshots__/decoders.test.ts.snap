// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decode should correctly decode parsed Demark documents 1`] = `
{
  "body": [
    {
      "paragraphs": [
        {
          "contents": [
            "I'm a very opinionated software developer. For example, I prefer writing ",
            {
              "tag": "code",
              "value": "for…of",
            },
            " loops over ",
            {
              "tag": "code",
              "value": ".forEach",
            },
            ". However, the Airbnb eslint config disallows ",
            {
              "tag": "code",
              "value": "for…of",
            },
            " loops. So, what's a developer supposed to do? Either you give in and use ",
            {
              "tag": "code",
              "value": ".forEach",
            },
            ", or you explicitly override the rule.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "Unfortunately, overriding a rule is not always as simple as turning it ",
            {
              "tag": "code",
              "value": ""off"",
            },
            ". For example, the Airbnb eslint config uses the ",
            {
              "href": "https://eslint.org/docs/latest/rules/no-restricted-syntax",
              "tag": "link",
              "value": {
                "tag": "code",
                "value": "no-restricted-syntax",
              },
            },
            " rule to disallow ",
            {
              "tag": "code",
              "value": "for…of",
            },
            " loops. If you turn it ",
            {
              "tag": "code",
              "value": ""off"",
            },
            " then you will inadvertently allow ",
            {
              "tag": "code",
              "value": "for…in",
            },
            " loops, labels, and the ",
            {
              "tag": "code",
              "value": "with",
            },
            " statement. Those are all things I want to disallow. Tweaking the rule to only allow ",
            {
              "tag": "code",
              "value": "for…of",
            },
            " loops is complicated.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "Now, though it might be complicated to tailor the Airbnb eslint config to suit your code style, it's nevertheless possible. But customized eslint configs suffer from another big problem — updates. Every time Airbnb updates its eslint config, there's a possibility that the updates might conflict with your code style. However, if you don't update then you skip out on configuring new rules.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "All these problems can be avoided if you create your own eslint config. Since you have full control of the eslint config, there's no possibility of subverting your code style. Furthermore, you can set up automated tests and workflows to ensure that your eslint config is always consistent and up to date. For example, the ",
            {
              "href": "https://github.com/AlloyTeam/eslint-config-alloy#philosophy",
              "tag": "link",
              "value": "Alloy",
            },
            " eslint config runs automated weekly checks for new rules added by eslint and its plugins. It also checks whether its rules conflict with ",
            {
              "href": "https://prettier.io/",
              "tag": "link",
              "value": "Prettier",
            },
            " or are deprecated.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "Sounds like a lot of work? It can indeed be daunting to create your own eslint config and manually configure all the rules to suit your code style. For example, I spent an entire 40-hour work week configuring all the rules in Ivory. But unlike me, you don't have to pay the cost upfront. You can create a fork of Ivory, start using it, and configure it over time to suit your code style.",
          ],
          "tag": "paragraph",
        },
      ],
      "tag": "section",
      "title": "1. You Will Have Full Control Over Your Code Style",
    },
    {
      "paragraphs": [
        {
          "contents": [
            "Creating your own eslint config does come at the significant cost of manually configuring all the rules. However, the benefit is that you only have to write the config once for all your repositories. For every new repository, you should only have to extend your eslint config. You should never have to override any rules, as that would make your code style inconsistent across repositories.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "I know from experience that it's very easy for developers to create multiple repositories with inconsistent code styles. It's effortless for developers to override rules on a per-repository basis. Especially when you're working in a large organization with multiple teams. Using an external eslint config like Airbnb just encourages developers to override rules on a per-repository basis.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "Now, inconsistent code styles would not be a major issue if eslint rules only enforced a stylistic preference, like ",
            {
              "tag": "code",
              "value": "for…of",
            },
            " loops vs ",
            {
              "tag": "code",
              "value": ".forEach",
            },
            ". However, a lot of eslint rules enforce a particular code style to prevent potential bugs. For example, I have a strong preference to ban type assertions in TypeScript by enabling the ",
            {
              "href": "https://typescript-eslint.io/rules/consistent-type-assertions",
              "tag": "link",
              "value": {
                "tag": "code",
                "value": "consistent-type-assertions",
              },
            },
            " rule and setting the ",
            {
              "tag": "code",
              "value": "assertionStyle",
            },
            " to ",
            {
              "tag": "code",
              "value": ""never"",
            },
            ". This prevents a lot of bugs by forcing developers to fix errors reported by the type checker. If this code style was inconsistent across repositories then bugs may slip through.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "When you create your own eslint config, you should discourage developers from overriding rules on a per-repository basis. Instead, you should encourage developers to raise an issue in your eslint config repository. This allows you to debate the proposed code style, update the eslint config in a disciplined manner, and most importantly maintain consistency across repositories.",
          ],
          "tag": "paragraph",
        },
      ],
      "tag": "section",
      "title": "2. You Will Have Consistency Across All Repositories",
    },
    {
      "paragraphs": [
        {
          "contents": [
            "While maintaining consistency across repositories is important, not all repositories require the same configuration rules. For example, you may only want to activate React rules for front-end applications. Fortunately, creating your own eslint config doesn't mean that you have to give up customization across repositories. You can have granular configs and combine them as required.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "Breaking big things into smaller things and then recombining them to form the original thing is a common theme in computer science. Programmers know this concept by many names such as ",
            {
              "href": "https://en.wikipedia.org/wiki/Dynamic_programming",
              "tag": "link",
              "value": "dynamic programming",
            },
            ", the ",
            {
              "href": "https://wiki.haskell.org/Combinator_pattern",
              "tag": "link",
              "value": "combinator pattern",
            },
            ", and ",
            {
              "href": "http://worrydream.com/LearnableProgramming/",
              "tag": "link",
              "value": "decomposition",
            },
            " and ",
            {
              "href": "https://www.cs.kent.ac.uk/people/staff/dat/miranda/whyfp90.pdf",
              "tag": "link",
              "value": "recomposition",
            },
            ". This technique has become popular in modern times. For example, the core concept of ",
            {
              "href": "https://tailwindcss.com/",
              "tag": "link",
              "value": "Tailwind CSS",
            },
            ", the most popular CSS framework, is combining simple style classes to build more complex styles.",
          ],
          "tag": "paragraph",
        },
        {
          "contents": [
            "The same concept can be applied to eslint configs. We already have a way of combining multiple eslint configs using the ",
            {
              "tag": "code",
              "value": "extends",
            },
            " property. So, instead of creating a single monolithic eslint config, we can create multiple eslint configs at the level of granularity we desire. We can then combine them using the ",
            {
              "tag": "code",
              "value": "extends",
            },
            " property into several preset configs for different kinds of repositories. The ",
            {
              "href": "https://github.com/gajus/eslint-config-canonical#example-configuration",
              "tag": "link",
              "value": "Canonical",
            },
            " eslint config has some excellent examples on how to combine multiple configs.",
          ],
          "tag": "paragraph",
        },
      ],
      "tag": "section",
      "title": "3. You Can Have Customization With Preset Configs",
    },
    {
      "paragraphs": [
        {
          "contents": [
            "Using a popular eslint config like Airbnb is easy when you want to get started as soon as possible. However, the return on investment sharply declines when you start overriding rules, copying your eslintrc file, and trying to maintain consistency across repositories. Having your own eslint config solves all these problems. And you don't need to start from scratch. Just fork ",
            {
              "href": "https://www.npmjs.com/package/eslint-config-ivory",
              "tag": "link",
              "value": "Ivory",
            },
            " and get going.",
          ],
          "tag": "paragraph",
        },
      ],
      "tag": "section",
      "title": "So Go Ahead and Create Your Own ESLint Config",
    },
  ],
  "introduction": [
    {
      "contents": [
        "I recently created my own eslint config called ",
        {
          "href": "https://www.npmjs.com/package/eslint-config-ivory",
          "tag": "link",
          "value": "Ivory",
        },
        ". It's my first open-source project. I decided that since I'm going to be a full-time open-source developer, I need to adhere to a single code style for all my repositories. So, why didn't I go with a popular eslint config like ",
        {
          "href": "https://github.com/airbnb/javascript",
          "tag": "link",
          "value": "Airbnb",
        },
        "? Here are 3 reasons why I created my own eslint config, and why you should consider doing the same.",
      ],
      "tag": "paragraph",
    },
  ],
  "tag": "article",
  "tags": [
    "eslint",
    "eslint-config",
    "ivory",
  ],
  "title": "3 Reasons Why You Should Create Your Own ESLint Config",
}
`;

exports[`decode should correctly decode parsed Demark documents with Unicode 1`] = `
{
  "content": [
    {
      "tag": "marker",
      "value": "",
    },
    {
      "tag": "marker",
      "value": "",
    },
    {
      "language": "demark",
      "tag": "code",
      "value": "
    \\article{Nested Document}~content #{
      \\paragraph{Demark documents support Unicode. For example, the value of π is approximately 3.141592653589793.}
      \\paragraph{You can also have emojis in your document. For example, here is a heart emoji ❤️.}
      \\paragraph{Demark documents also support characters outside the Basic Multilingual Plane such as this fancy 𝕽.}
    #}
  ",
    },
  ],
  "tag": "article",
  "title": "Test Document",
}
`;
